zenoh/src/net/routing/hat/client/token.rs

```diff
@@ -0,0 +1,383 @@
+        if let Some(res) = res.as_ref() {
+            if aggregate {
+                if tables.faces.values().any(|src_face| {
+                    face_hat!(src_face)
+                        .remote_tokens
+                        .values()
+                        .any(|token| token.context.is_some() && token.matches(res))
+                }) {
+                    let id = if mode.future() {
+                        let id = face_hat!(face).next_id.fetch_add(1, Ordering::SeqCst);
+                        face_hat_mut!(face).local_tokens.insert((*res).clone(), id);
+                        id
+                    } else {
+                        0
+                    };
+                    let wire_expr = Resource::decl_key(res, face);
+                    face.primitives.send_declare(RoutingContext::with_expr(
+                        Declare {
+                            interest_id,
+                            ext_qos: ext::QoSType::DECLARE,
+                            ext_tstamp: None,
+                            ext_nodeid: ext::NodeIdType::DEFAULT,
+                            body: DeclareBody::DeclareToken(DeclareToken { id, wire_expr }),
+                        },
+                        res.expr(),
+                    ));
+                }
+            } else {
+                for src_face in tables
+                    .faces
+                    .values()
+                    .cloned()
+                    .collect::<Vec<Arc<FaceState>>>()
+                {
+                    for token in face_hat!(src_face).remote_tokens.values() {
+                        if token.context.is_some() && token.matches(res) {
+                            let id = if mode.future() {
+                                let id = face_hat!(face).next_id.fetch_add(1, Ordering::SeqCst);
+                                face_hat_mut!(face).local_tokens.insert(token.clone(), id);
+                                id
+                            } else {
+                                0
+                            };
+                            let wire_expr = Resource::decl_key(token, face);
+                            face.primitives.send_declare(RoutingContext::with_expr(
+                                Declare {
+                                    interest_id,
+                                    ext_qos: ext::QoSType::default(),
+                                    ext_tstamp: None,
+                                    ext_nodeid: ext::NodeIdType::default(),
+                                    body: DeclareBody::DeclareToken(DeclareToken { id, wire_expr }),
+                                },
+                                res.expr(),
+                            ))
+                        }
+                    }
+                }
+            }
+        } else {
+            for src_face in tables
+                .faces
+                .values()
+                .cloned()
+                .collect::<Vec<Arc<FaceState>>>()
```

Comment:
Is `collect` really needed here?

Reply:
Yes. Otherwise there would be a double borrow on tables.